### PR TITLE
fix stack overflow on deeply nested parens [SR-4866]

### PR DESCRIPTION
**Pull request by modelorganism**
Currently swiftc crashes with no diagnostic on deeply nested parens, but issues a diagnostic for square brackets nested more than 256 deep. 

This PR applies the limit in all cases where we push onto the StructureMarker stack.Now we print an error when more than 256 parens deep, just as we always have for '['.

This change alters the `swift::Parser::StructureMarkerRAII` constructors to call a common 
function `pushStructureMarker` where we check structure nesting level and call a halt to parsing
if we are too deep. Previously structure nesting level limits were enforced in some cases,
but not others. Most of the files touched in this PR are simply to add a check and early 
return if we have hit the structure nesting limit and called a halt to parsing.

Note that even with this PR swift will in some cases hit an assertion failure after we issue the diagnostic, but this is better than the current crash w/o any diagnostic.
Fixing this remaining issue would require more extensive changes to the parser.

Improves [SR-4866](https://bugs.swift.org/browse/SR-4866).
Though the situation is still not perfect.

### CODE REVIEW

**Comment by modelorganism** [comment]
@harlanhaskins thanks for looking at this,
I've put one test back,  it dose _not_ succeed  due to a 
<unknown>:0: error: fatal error encountered while in -verify mode
but I'll try to debug it

**Comment by rintaro** [comment]
Thanks for fixing this!
Fatal errors cannot be verified with `-verify`.
`test/Parse/structure_overflow.swift` is a good example for testing it.

**Comment by modelorganism** [comment]
@rintaro yes, thank you, the method from test/Parse/structure_overflow.swift dose work.  So tests are in and working now. One test that we can approach the limit w/o error, and the original example from bug SR-4866 test-ified that we get a diagnostic, not a crash.  I as far as I know we are good to go, let me know if there is anything else needed.

File: lib/Parse/Parser.cpp [hunk 1]
```diff
@@ -757,38 +757,27 @@ bool Parser::parseEndIfDirective(SourceLoc &Loc) {
...
   }
 }
 
-void Parser::StructureMarkerRAII::diagnoseOverflow() {
-  auto Loc = P.StructureMarkers.back().Loc;
-  P.diagnose(Loc, diag::structure_overflow, MaxDepth);
-}
+
+Parser::StructureMarkerRAII::StructureMarkerRAII(Parser &parser, const Token &tok):
+  StructureMarkerRAII(parser, tok.getLoc(), getStructureMarkerKindForToken(tok)) { }
```
Comment (rintaro) [inline lib/Parse/Parser.cpp]: 80 columns.

File: lib/Parse/Parser.cpp [hunk 2]
```diff
@@ -757,38 +757,27 @@ bool Parser::parseEndIfDirective(SourceLoc &Loc) {
   return false;
 }
 
-Parser::StructureMarkerRAII::StructureMarkerRAII(Parser &parser,
-                                                 const Token &tok)
-  : P(parser)
+static Parser::StructureMarkerKind getStructureMarkerKindForToken(const Token &tok)
```
Comment (rintaro) [inline lib/Parse/Parser.cpp]: 80 columns

File: include/swift/Parse/Parser.h [hunk 3]
```diff
@@ -306,30 +306,47 @@ class Parser {
...
-      if (P.StructureMarkers.size() >= MaxDepth) {
-        diagnoseOverflow();
-        P.cutOffParsing();
-      }
+      
+  StructureMarkerRAII(Parser *parser)
+      : P(parser) {  }
+    
+  // Have the parser start the new Structure or fail if already too deep.
+  bool pushStructureMarker(Parser &parser, SourceLoc loc, StructureMarkerKind kind) {
```
[the remark this answers is not in the document]
Comment (rintaro) [reply include/swift/Parse/Parser.h]: Also, could you put this method to `Parser.cpp`?

**Review by rintaro — COMMENTED** [review]
Thanks! Mostly coding style requests.

Just FYI, `git-clang-format` is a good tool to comply with our coding style.

File: test/Parse/structure_no_overflow.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
+// RUN: %target-typecheck-verify-swift
+
+// Test that we can have the allowed # of parens.
+// CHECK-NOT: error
```
[the remark this answers is not in the document]
Comment (modelorganism) [reply test/Parse/structure_no_overflow.swift]: I see, %FileCheck has been added

File: lib/Parse/Parser.cpp (hunk 1 above)
Comment (modelorganism) [reply lib/Parse/Parser.cpp]: `80+-check` now reports now reports no violations in the parts I touched.

File: include/swift/Parse/Parser.h [hunk 5]
```diff
@@ -306,30 +306,47 @@ class Parser {
...
-                               StructureMarkerKind kind)
-      : P(parser) {
-      P.StructureMarkers.push_back({loc, kind, None});
-
-      if (P.StructureMarkers.size() >= MaxDepth) {
-        diagnoseOverflow();
-        P.cutOffParsing();
-      }
+      
+  StructureMarkerRAII(Parser *parser)
```
[the remark this answers is not in the document]
Comment (modelorganism) [reply include/swift/Parse/Parser.h]: done

File: test/Parse/structure_no_overflow.swift [hunk 6]
```diff
@@ -0,0 +1,27 @@
+// RUN: %target-typecheck-verify-swift
```
[the remark this answers is not in the document]
Comment (modelorganism) [reply test/Parse/structure_no_overflow.swift]: switched to -parse, 200 '['s was still over 20 min, but I put in a mixed case with some '['.

File: include/swift/Parse/Parser.h [hunk 7]
```diff
@@ -306,30 +306,47 @@ class Parser {
...
+      // and we may be in a place where `consumeToken()` will be expecting e.g. '[',
+      // since we need that to get to the callsite, so this can cause an assert.
+      parser.cutOffParsing();
+      return false;
     }
+  }
 
+  public:
+    StructureMarkerRAII(Parser &parser, SourceLoc loc, StructureMarkerKind kind):
+      StructureMarkerRAII( pushStructureMarker(parser, loc, kind)? &parser : nullptr ) {}
```
[the remark this answers is not in the document]
Comment (modelorganism) [reply include/swift/Parse/Parser.h]: ok

File: include/swift/Parse/Parser.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (modelorganism) [reply include/swift/Parse/Parser.h]: ok moved

File: lib/Parse/Parser.cpp (hunk 2 above)
Comment (modelorganism) [reply lib/Parse/Parser.cpp]: ok

**Comment by modelorganism** [comment]
Everything I'm aware of is in.

This PR Resolves SR-4866.
(It's possible there are still undiscovered asserts that would fire after the cutoff of parsing,
bit this fixes the problem described. So not calling it resolved my be unnecessarily  pessimistic).

File: lib/Parse/Parser.cpp [hunk 8]
```diff
@@ -757,39 +757,42 @@ bool Parser::parseEndIfDirective(SourceLoc &Loc) {
...
-    break;
-
+    return Parser::StructureMarkerKind::OpenBrace;
   case tok::l_paren:
-    P.StructureMarkers.push_back({tok.getLoc(),
-                                  StructureMarkerKind::OpenParen,
-                                  None});
-    break;
-
+    return Parser::StructureMarkerKind::OpenBrace;
```
Comment (rintaro) [inline lib/Parse/Parser.cpp]: `OpenBrace` -> `OpenParen`

File: lib/Parse/ParseType.cpp [hunk 9]
```diff
@@ -852,6 +852,12 @@ ParserResult<TypeRepr> Parser::parseOldStyleProtocolComposition() {
 ParserResult<TupleTypeRepr> Parser::parseTypeTupleBody() {
   SyntaxParsingContext TypeContext(SyntaxContext, SyntaxKind::TupleType);
   Parser::StructureMarkerRAII ParsingTypeTuple(*this, Tok);
+   
+  // consumeToken would panic if it saw the eof from StructureMarkerRAII halting the parse.
```
Comment (rintaro) [inline lib/Parse/ParseType.cpp]: 80columns.
But I don't think we need this comment.

File: test/Parse/structure_no_overflow.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (rintaro) [reply test/Parse/structure_no_overflow.swift]: > switched to -parse, 200 '['s was still over 20 min,
Confirmed. I'll look into it.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]

**Comment by rintaro** [comment]
Thank you Christopher!

**Pull request merged**